File: app/src/main/java/com/mikhailovskii/trakttv/ui/favorites/FavoritesFragment.java
package com.mikhailovskii.trakttv.ui.favorites;

import android.content.Intent;
import android.os.Bundle;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.mikhailovskii.trakttv.R;
import com.mikhailovskii.trakttv.data.entity.Movie;
import com.mikhailovskii.trakttv.ui.adapter.MoviesAdapter;
import com.mikhailovskii.trakttv.ui.movie_detail.MovieDetailActivity;
import com.mikhailovskii.trakttv.ui.movies_list.MovieListFragment;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Objects;

public class FavoritesFragment extends Fragment
        implements FavoritesContract.FavoritesView, MoviesAdapter.OnItemClickListener {

    private FavoritesPresenter mPresenter = new FavoritesPresenter();
    private TextView mTvNoFilms;
    private SwipeRefreshLayout mSwipeRefresh;
    private RecyclerView mMoviesRecycler;
    private MoviesAdapter mAdapter;

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_favorites, container, false);
        mPresenter.attachView(this);

        mTvNoFilms = view.findViewById(R.id.no_films);
        mSwipeRefresh = view.findViewById(R.id.swipe_refresh);
        mSwipeRefresh.setOnRefreshListener(() -> mPresenter.loadFavoriteMovies());

        mMoviesRecycler = view.findViewById(R.id.movies_list);
        mMoviesRecycler.setLayoutManager(new LinearLayoutManager(getContext()));
        mMoviesRecycler.addItemDecoration(new DividerItemDecoration(Objects.requireNonNull(getActivity()), DividerItemDecoration.VERTICAL));
        mAdapter = new MoviesAdapter(this);
        mMoviesRecycler.setAdapter(mAdapter);

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        mPresenter.loadFavoriteMovies();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.detachView();
    }

    @Override
    public void onMoviesLoaded(List<Movie> list) {
        mAdapter.setData(list);
    }

    @Override
    public void onMoviesFailed() {
        Toast.makeText(getContext(), getString(R.string.loading_failed), Toast.LENGTH_SHORT).show();
        showEmptyState(true);
    }

    @Override
    public void onMovieDeleted() {
        Toast.makeText(getContext(), "Movie deleted", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onMovieDeleteFailed() {

    }

    @Override
    public void showEmptyState(@NonNull Boolean value) {
        if (value) {
            mTvNoFilms.setVisibility(View.VISIBLE);
        } else {
            mTvNoFilms.setVisibility(View.GONE);
        }
    }

    @Override
    public void showLoadingIndicator(@NonNull Boolean value) {
        mSwipeRefresh.setRefreshing(value);
    }

    @Override
    public void onItemClicked(int position, Movie item) {
        Intent intent = new Intent(getActivity(), MovieDetailActivity.class);
        intent.putExtra(MovieListFragment.EXTRA_IMAGE, item.getIconUrl());
        intent.putExtra(MovieListFragment.EXTRA_SLUG, item.getSlugId());
        startActivity(intent);
    }

    @Override
    public void onItemLongClick(int position, Movie item) {
        Bundle bundle = new Bundle();
        bundle.putString(MovieDetailActivity.EXTRA_NAME, item.getName());
        bundle.putInt(MovieDetailActivity.EXTRA_WATCHERS, item.getWatchers());
        bundle.putString(MovieListFragment.EXTRA_SLUG, item.getSlugId());
        bundle.putString(MovieListFragment.EXTRA_IMAGE, item.getIconUrl());
        mPresenter.deleteMovie(bundle);
    }

}

File: app/src/main/java/com/mikhailovskii/trakttv/ui/favorites/FavoritesFragment.kt
package com.mikhailovskii.trakttv.ui.favorites

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.GridLayoutManager
import com.mikhailovskii.trakttv.R
import com.mikhailovskii.trakttv.data.diffutil.MovieDiffUtilCallback
import com.mikhailovskii.trakttv.data.entity.Movie
import com.mikhailovskii.trakttv.ui.adapter.MoviesAdapter
import com.mikhailovskii.trakttv.ui.movie_detail.MovieDetailActivity
import com.mikhailovskii.trakttv.ui.movies_list.MovieListFragment
import com.mikhailovskii.trakttv.util.errorToast
import com.mikhailovskii.trakttv.util.successToast
import kotlinx.android.synthetic.main.fragment_favorites.*
import org.koin.android.scope.currentScope
import java.util.*

class FavoritesFragment : Fragment(), FavoritesContract.FavoritesView, MoviesAdapter.OnItemClickListener {

    private val presenter by currentScope.inject<FavoritesContract.FavoritesPresenter>()

    private var adapter: MoviesAdapter? = null

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_favorites, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        presenter.attachView(this)

        swipe_refresh.setOnRefreshListener {
            presenter.loadFavoriteMovies()
        }

        movies_list.layoutManager = GridLayoutManager(context, resources.getInteger(R.integer.rows))
        movies_list.addItemDecoration(DividerItemDecoration(Objects.requireNonNull<FragmentActivity>(activity), DividerItemDecoration.VERTICAL))
        adapter = MoviesAdapter(this)
        movies_list.adapter = adapter

    }

    override fun onResume() {
        super.onResume()
        presenter.loadFavoriteMovies()
    }

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.detachView()
    }

    override fun onMoviesLoaded(list: List<Movie>) {

        val movieDiffUtilCallback = MovieDiffUtilCallback(list, adapter?.movieList!!)
        val movieDiffResult: DiffUtil.DiffResult = DiffUtil.calculateDiff(movieDiffUtilCallback)
        adapter?.setData(list)
        adapter?.let {
            movieDiffResult.dispatchUpdatesTo(it)
        }
    }

    override fun onMoviesFailed() {
        errorToast(getString(R.string.loading_failed))
    }

    override fun onMovieRemoved() {
        successToast(getString(R.string.movie_removed))
        presenter.loadFavoriteMovies()
    }

    override fun onMovieRemoveFailed() {
        errorToast(getString(R.string.movie_remove_failed))

    }

    override fun showEmptyState(value: Boolean) {
        if (value) {
            tv_no_films.visibility = View.VISIBLE
            movies_list.visibility = View.GONE
        } else {
            tv_no_films.visibility = View.GONE
            movies_list.visibility = View.VISIBLE
        }
    }

    override fun showLoadingIndicator(value: Boolean) {
        swipe_refresh.isRefreshing = value
    }

    override fun onItemClicked(position: Int, item: Movie) {
        val intent = Intent(activity, MovieDetailActivity::class.java)
        intent.putExtra(MovieListFragment.EXTRA_IMAGE, item.iconUrl)
        intent.putExtra(MovieListFragment.EXTRA_SLUG, item.movieId?.slug)
        startActivity(intent)
    }

    override fun onItemLongClick(position: Int, item: Movie) {
        presenter.deleteMovie(item.name!!)
    }

    companion object {

        fun newInstance(): FavoritesFragment = FavoritesFragment()

    }

}

